File: Solutions/209. Minimum Size Subarray Sum/TypeScript/Minimum_Size_Subarray_Sum.ts
function minSubArrayLen(target: number, nums: number[]): number {
  let sum = 0;
  let min_length = 0;

  let i = 0, j = 0;
  while (j < nums.length || sum > target) {
    if (sum < target) {
      sum += nums[j];
      j++;
    } else if (sum >= target) {
      sum -= nums[i];
      i++;
    }

    if (sum >= target) {
      min_length = (min_length == 0) ? j - i : Math.min(min_length, j - i);
    }
  }

  return min_length;
};

const target = 7;
const nums = [2, 3, 1, 2, 4, 3];

console.log(minSubArrayLen(target, nums));

File: Solutions/209. Minimum Size Subarray Sum/CPP/Minimum_Size_Subarray_Sum.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution{
public:
    int minSubArrayLen(int target, vector<int>& nums) {
        int sum = 0;
        int min_length = 0;

        int i = 0, j = 0;
        while (j < nums.size() || sum > target) {
            if (sum <= target) {
                sum += nums[j];
                j++;
            }
            else if (sum > target) {
                sum -= nums[i];
                i++;
            }

            if (sum >= target) {
                min_length = (min_length == 0) ? j - i : min(min_length, j - i);
            }
        }

        return min_length;
    }
};

int main(){
    Solution *S = new Solution();

    //inputs
    int target = 7;
    vector<int> nums = {2, 3, 1, 2, 4, 3};

    cout << S->minSubArrayLen(target, nums) << endl;

    return 0;
}